_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,263 @@
+---
+layout: post
+date: 2024-02-06
+title: "Build cache in EO and other build systems"
+author: Alekseeva Yana
+---
+
+
+## Introduction 
+In [EO](https://github.com/objectionary/eo), caching is used to speed up program compilation.
+Recently we found a caching 
+[bug](https://github.com/objectionary/eo/issues/2790) in `eo-maven-plugin`
+for EO version `0.34.0`. The bug occurred because the old verification method
+contains a comparison of the compilation time and caching time to search for the cached file.
+This is not the most reliable verification method,
+because caching time does not have to be equal to compilation time.
+We came to the conclusion that we need caching with a reliable verification method.
+And this verification method should not use the information that the cached file contains.
+
+The goal of this blog is to research caching in frequently used build systems (`ccache`, `Maven`, `Gradle`)
+and to implement effective caching in [EO](https://github.com/objectionary/eo).
+
+<!--more-->
+
+## Build caching of existing build systems
+
+### ccache/sccache
+In compiled programming languages, building a project containing many source code files takes a long time.
+This time is spent on loading of libraries, preparing, optimizing, checking the code, and so on.
+To speed up the assembly of compiled languages, [ccache](https://ccache.dev) 
+and [sccache](https://github.com/mozilla/sccache) are used.
+Let's look at the assembly scheme using C++ as an example
```

Comment:
@Yanich96 To be honest, I have some doubts about this paragraph where you discuss "compilation steps":
1) First of all this is a blog about caching, not about compilation
2) You describe compilation incompletely. What about optimizations? Moreover, modern compilers usually convert source code to some sort of IR, like LLVM IR, for example. You can take a look how [clang](https://en.wikipedia.org/wiki/Parsing) works.

Reply:
@volodya-lombrozo I believe that a reader will better understand this article if we briefly talk about the stages of compilation of the presented build systems. Yes, I describe compilation incompletely, but enough to indicate where caching works.

Reply:
@Yanich96 Then, it's good to mention it.

Reply:
@volodya-lombrozo "The goal is to implement effective caching in [EO](https://github.com/objectionary/eo).
For this, we will briefly look at how frequently used build systems (`ccache`, `Maven`, `Gradle`) work
in order to better understand the ideas behind caching in them."
 it's ok?